Reformat Transform(2D) matrix display in the inspector

aaronfranke opened the pull request:
Previously the Transform editors in the inspector were very sub-optimal. Columns were displayed horizontally, rows displayed vertically, there was repeated information, and labels were non-unique.

This PR fixes all of these problems. Here are some before and after screenshots, the left side is before, the right side is after.

![reformat-transform-new-version](https://user-images.githubusercontent.com/1646875/131546222-f7736027-b70d-4d0f-bf0a-4feef2ef88f8.png)


The items have been transposed to display columns vertically and rows horizontally (as it should be). I decided to use the fourth color as Cyan, which I think looks better. This Cyan color also replaces gray in `Quaternion` and `Plane` (but of course we can change this back if desired).

The names are row-column, just like in math, such that "yx" is equivalent to `transform.basis.x.y`, etc. This is much better than the previous ambiguous, repeated, non-unique "x", "y", and "z". This notation was suggested by tagcup.

Coloring is now done by a helper method in `editor/editor_inspector.cpp`. I also made everything else use these helper methods, which helps avoid code duplication.

<details> 

EDIT: Marked as draft because #47635 should be merged before this one (it's a subset).

Implements and closes https://github.com/godotengine/godot-proposals/issues/1820, and fixes #44669.

Previously the Transform editors in the inspector were a bit of a mess. Columns were displayed horizontally, rows displayed vertically, there was repeated information, and labels were non-unique. Also, Node2D is completely missing a section for its matrix, so it's impossible to edit a 2D node's transform matrix directly without using a script.

This PR fixes all of these problems. Here are some before and after screenshots, the left side is before, the right side is after. There are exported values, Node2D, and Node3D. Each side has equivalent values to each other, the Y component of the X column vector (row 2 col 1) is set to 5 for most of the examples, and for Node2D the rotation is set to 5.

![Untitled](https://user-images.githubusercontent.com/1646875/111742689-fd207c00-885e-11eb-806b-86fd52875422.png)

The "Matrix" section has been renamed to "Raw Matrix" to better convey that this is how Godot works under-the-hood, and to slightly discourage users who don't know what it is from touching it. Editing the matrix of a transform directly is something users should know how it works before jumping in, or else they will get confused.

The items have been transposed to display columns vertically and rows horizontally (as it should be), except that for Node2D and Node3D, the origin vector is not displayed, instead, only the Basis is displayed. This is because we don't need to duplicate information already present in "Translation". The origin vector is still displayed for exported transforms. I decided to color it Cyan, rather than leave it gray, which I think looks better. This Cyan color also replaces gray in `Quat` and `Plane` (but of course we can change this back if desired).

The names are row-column, just like in math, such that "yx" is equivalent to `transform.basis.x.y`, etc. This is much better than the previous ambiguous, repeated, non-unique "x", "y", and "z".

Previously, the transform matrix was not displayed at all for Node2D. The display now exists, and closely resembles the section for Node3D, it only shows the basis part (excludes the origin vector) because the information exists in "Position".

Coloring is now done by a helper method in `editor/editor_inspector.cpp`. I also made everything else use these helpers methods, which helps avoid code duplication (and also, while I haven't benchmarked it, I'm fairly confident this will run faster since it's not re-calculating the same colors multiple times anywhere that uses the colors multiple times).

tagcup commented:
transform.basis.x.y would be yx not xy

tagcup commented:
Row is the first letter for matrices in standard math notation

aaronfranke commented:
@tagcup I think it's more intuitive conceptually to think of a Transform of its set of component vectors (columns) and their components, and that is how programmers would be interacting with transform most of the time, so the `xy` == `x.y` makes more sense for Godot. We could also just add a `.` between the letters to show this if desired.

tagcup commented:
I disagree. That's against every math or physics book out there so we shouldn't redefine standard linear algebra notations, or introduce new notations (the dotted version which becomes one new nonstandard thing that you need to explain) which flip the standards upside down and would ultimately create confusion only down the road.

I understand the urge stems from identifying the columns of the transform matrix as basis vectors of the frame, but everything actually reads correct if you use the correct, standard notation, and would be wrong, inconsistent and ultimately confusing with that notation.

The reason is when you multiply a matrix with a vector, you sum over the second index not the first one: R_ij v_j The basis vector interpretation is when v is (1, 0, 0), (0, 1, 0) or (0, 0, 1) which gives you (Rxx, Ryx, Rzx) and so on, as it should. If R.(1,0,0) was (Rxx, Rxy, Rxz) it would just be wrong.

It's be to accept as a matrix for what is it: a matrix (or a rank 2 tensor in this case), and avoid such confusing notations that flip the literature upside down.

aaronfranke commented:
Should the origin column be changed to `xo`, `yo`, and `zo` in your opinion?

I still think it's best to reflect how the matrix is actually used in Godot, so I'm still in favor of the column-major syntax, but I will add dots (unless we reach a consensus that we should indeed use the row-major syntax). This isn't just any matrix, it's a specific kind of matrix of a fixed size and its components have specific purposes. It would be confusing if you accessed `xy` with `y.x`.

tagcup commented:
- There is no column major in maths or physics, that's just a computer language convention in languages which support multi dimensional arrays related memory access speed. A matrix that is printed on the screen must always follow the row major. See the image here:

https://en.m.wikipedia.org/wiki/Row-_and_column-major_order


- C++ basis in Godot is actually row major b[row][column]. Transform 2D is column major for whatever reason

- Vectors are columns, that's consistent with typical math on paper. That has nothing to do with row or column major

- Old OpenGL transposed everything for some obsolete optimization (which caused a lot confusion because you had to transpose everything you feed, they now have a flag for this behavior) so that matrix multiplication with a vector is vi Rij. Godot math library doesn't do that, R.v is Rij vj also consistent with math on paper so it's not relevant either

aaronfranke commented:
@tagcup I've added the ability to display matrices with math notation as a 2nd commit:

![Use Math Notation For Matrices Checkbox](https://user-images.githubusercontent.com/1646875/66863132-5dbbd300-ef60-11e9-9591-e13b42bfe02a.png)

When enabled, the inspector looks like this:

![Matrix Display Transform Transform2D Basis](https://user-images.githubusercontent.com/1646875/66863142-601e2d00-ef60-11e9-9125-f9a201c5ebf5.png)

The "M" could either mean math or matrix, and in any case it's useful to differentiate it in case users are creating screenshots/videos/tutorials and are wondering why the labels look different.

It's disabled by default right now, but it would be trivial to change this if you wish to open a new issue after this pull request is merged (assuming it will be).

> the basis vector interpretation is already conveyed through the colors

Yes, but note that this isn't the case in the current master.

> The translation vector in the "raw matrix" is redundant, the same vector is displayed twice

Yes, which is why I made Spatial display the Basis instead. However, there is no separate Basis2D type, so we can't do the same for Node2D unless we want to implement Basis2D, but I think that's out of the scope for this pull request. Or maybe there's a workaround I haven't considered.

tagcup commented:
For the reasons that I explained above multiple times, Basis should never be displayed like that, having such modes is even more confusing. No such notation exists anywhere, and Godot should never have it.  This should never be merged.

Basis is a matrix, not an array of vectors. I know it is tempting to try to interpret it like that but there are very different things. Just accept the notations the mathematicians and physicists and everyone else agreed on and please stop pushing this nonsense any further.

Please don't ping me anymore. I'm 100% against this in any form, optional or not, I gave the reason, and I don't have anything else to add.

aaronfranke commented:
Dug into the code a bit further, figured out a way to make Node2D's raw matrix display without the origin. It checks if `p_usage` is 2 or not, so it may unexpectedly hide Transform2D if `p_usage` is 2 in other places with Transform2D (I don't think there are any, though?). Also, it still says "Transform".

![node2d-transform2d](https://user-images.githubusercontent.com/1646875/70515759-d758e300-1b03-11ea-94ea-eb914ddff7d4.png)

fire commented:
I don't understand the benefit of this.

aaronfranke commented:
@fire Read the OP. In master, these editors are very confusing. Columns are displayed horizontally, rows displayed vertically, there was repeated information, and labels were non-unique. It doesn't make sense to display rows vertically and it doesn't make sense to display columns horizontally.

aaronfranke committed:
Reformat Transform(2D) inspector menus

Move, rename, recolor, and fix the inspector for these in the inspector. Coloring is now done by a helper method in "editor_inspector.cpp".

Calinou commented on editor/editor_properties.cpp:
```diff
@@ -2060,7 +2052,12 @@ EditorPropertyBasis::EditorPropertyBasis() {
-	static const char *desc[9] = { "x", "y", "z", "x", "y", "z", "x", "y", "z" };
+	static const char **desc;
+	if (EDITOR_GET("interface/inspector/use_vector_component_notation_for_matrices")) {
+		desc = new const char *[9] { "x.x", "y.x", "z.x", "x.y", "y.y", "z.y", "x.z", "y.z", "z.z" };
```
Shouldn't we use `Array()`/`PackedStringArray()` instead of C-style arrays created with `new` and `delete`?

Alternatively, you can remove the editor setting since it's disabled by default. I wouldn't add an editor setting for something that doesn't have significant demand (i.e. don't future-proof settings).

aaronfranke replied:
I don't know how to create a PackedStringArray with values already in it.

Anyway, I removed the setting, you're right that it's not necessary.

Calinou reviewed (commented):
The new inspectors look better and more usable, good work on that :slightly_smiling_face: 

The code seems alright to me, but I'm not sure about adding yet another editor setting to an already cluttered Editor Settings dialog.

fire commented:
Can we split the matrix convention change from the current pr?

I can add my approval for the rest of the changes.

However, I'm not sure if transposing the matrix is better. My opinion is that you're changing conventions for people and people need to relearn. 

I'm not convinced by the convention change. However that shouldn't block the rest of the changes.

![image](https://user-images.githubusercontent.com/32321/113168788-d61c5f80-91f9-11eb-9e3b-9993b1a3014a.png)

reduz commented:
I dont kinda like the idea of changing this as I prefer it the current way, which works better for horizontally compact inspectors.
Also, #44669 is no longer a bug in master.

aaronfranke commented:
@reduz There are many problems with the current master, so we should still do something, but we can change this PR. If being horizontally compact is important, we could change it so that the origin vector is at the bottom like it is now, but keep the other changes in this PR including making rows display horizontally as rows and (non-origin) columns display vertically as columns.

(Note that the placement of the origin vector will only affect exported variables, since for the Node2D and Node3D inspectors the origin vector is hidden as it's not needed)

aaronfranke commented:
@reduz Is this better? I added this change as a 2nd commit.

![Screenshot from 2021-04-04 22-04-43](https://user-images.githubusercontent.com/1646875/113529647-f5b4de80-9591-11eb-9bed-e5c8beba0c43.png)

reduz commented:
I think as far as showing transforms in itself, this is ok and should be able to go through. About the position I believe there is a different PR/Idea for this.

Calinou commented:
Is there a reason Transform2D's `xo` and `yo` labels are blue (like the Z axis) instead of cyan? I'd expect the Transform origin properties to always be cyan, regardless if they're 2D or 3D.

aaronfranke commented:
@Calinou Good idea, done, and updated the image.

akien-mga commented:
Thanks!
